AAP-39365 facts are unintentionally deleted when the inventory is modified during a job execution

djyasin opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This change fixes facts are unintentionally deleted when the inventory is modified during a job execution
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 24.6.2
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

onetti7 committed:
Added test_jobs.py to the model unit test folder in orther to show the undesired behaviour with fact cache

Signed-off-by: onetti7 <davonebat@gmail.com>

onetti7 committed:
Added test_jobs.py to the model unit test folder in orther to show the undesired behaviour with fact cache

Signed-off-by: onetti7 <davonebat@gmail.com>

onetti7 committed:
Solved undesired behaviour with fact_cache

Signed-off-by: onetti7 <davonebat@gmail.com>

onetti7 committed:
Solved bug with slices

Signed-off-by: onetti7 <davonebat@gmail.com>

djyasin committed:
Remove unused imports

Remove now unused line of code which was commented out by the contributor

Revert "Remove now unused line of code which was commented out by the contributor"

This reverts commit f1a056a2356d56bc7256957a18503bd14dcfd8aa.

djyasin committed:
Add back line that had been commented out as this line makes hosts specific to the particular slice when applicable

Revise private_data_dir fixture to see if it improves code coverage

Checked out awx/main/tests/unit/models/test_jobs.py in devel to see if it resolves git diff issue

AlanCoding commented on awx/main/tests/unit/models/test_jobs.py:
```diff
@@ -1,142 +1,142 @@
+        assert host.ansible_facts_modified == ref_time
+    assert hosts[1].ansible_facts == {}
+    assert hosts[1].ansible_facts_modified > ref_time
+    bulk_update.assert_called_once_with([hosts[1]], ['ansible_facts', 'ansible_facts_modified'])
```
Something seems strange with the git history here. Why does it show the entire file as changed? Manually look through the first lines, it's not. I'm worried that there is some git corruption in this branch.

AlanCoding replied:
This is the diff it should be showing

```diff
diff --git a/awx/main/tests/unit/models/test_jobs.py b/awx/main/tests/unit/models/test_jobs.py
index 4f05a82535..f35b633f76 100644
--- a/awx/main/tests/unit/models/test_jobs.py
+++ b/awx/main/tests/unit/models/test_jobs.py
@@ -34,7 +34,7 @@ def hosts(ref_time):
 
 def test_start_job_fact_cache(hosts, tmpdir):
     fact_cache = os.path.join(tmpdir, 'facts')
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=0)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=0)
 
     for host in hosts:
         filepath = os.path.join(fact_cache, host.name)
@@ -61,7 +61,7 @@ def test_fact_cache_with_invalid_path_traversal(tmpdir):
 def test_start_job_fact_cache_past_timeout(hosts, tmpdir):
     fact_cache = os.path.join(tmpdir, 'facts')
     # the hosts fixture was modified 5s ago, which is more than 2s
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=2)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=2)
     assert last_modified is None
 
     for host in hosts:
@@ -71,7 +71,7 @@ def test_start_job_fact_cache_past_timeout(hosts, tmpdir):
 def test_start_job_fact_cache_within_timeout(hosts, tmpdir):
     fact_cache = os.path.join(tmpdir, 'facts')
     # the hosts fixture was modified 5s ago, which is less than 7s
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=7)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=7)
     assert last_modified
 
     for host in hosts:
@@ -80,7 +80,7 @@ def test_start_job_fact_cache_within_timeout(hosts, tmpdir):
 
 def test_finish_job_fact_cache_with_existing_data(hosts, mocker, tmpdir, ref_time):
     fact_cache = os.path.join(tmpdir, 'facts')
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=0)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=0)
 
     bulk_update = mocker.patch('django.db.models.query.QuerySet.bulk_update')
 
@@ -108,7 +108,7 @@ def test_finish_job_fact_cache_with_existing_data(hosts, mocker, tmpdir, ref_tim
 
 def test_finish_job_fact_cache_with_bad_data(hosts, mocker, tmpdir):
     fact_cache = os.path.join(tmpdir, 'facts')
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=0)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=0)
 
     bulk_update = mocker.patch('django.db.models.query.QuerySet.bulk_update')
 
@@ -127,7 +127,7 @@ def test_finish_job_fact_cache_with_bad_data(hosts, mocker, tmpdir):
 
 def test_finish_job_fact_cache_clear(hosts, mocker, ref_time, tmpdir):
     fact_cache = os.path.join(tmpdir, 'facts')
-    last_modified = start_fact_cache(hosts, fact_cache, timeout=0)
+    last_modified, _ = start_fact_cache(hosts, fact_cache, timeout=0)
 
     bulk_update = mocker.patch('django.db.models.query.QuerySet.bulk_update')
```

The fact that it's not showing this diff is suggesting that something is messed up.

djyasin replied:
Uh-oh- Okay, @AlanCoding thank you! I am going to make a test PR and re-pick the commits to see if it shows up correctly. I will go do some digging around that. Thank you!

djyasin replied:
@AlanCoding I believe this has been fixed! Thank you for your help!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -130,6 +131,9 @@ def finish_fact_cache(hosts, destination, facts_write_time, log_data, job_id=Non
             # if the file goes missing, ansible removed it (likely via clear_facts)
+            # if the file goes missing, but the host hasnot started facts, then we should not clear the facts
+            if hosts_cached and host.name not in hosts_cached:
+                continue
```
The outer iteration loop no longer makes any sense, that being `for host in hosts:`. On the surface-level you might as well change that to `for host in hosts_cached:` (and any associated changes), and get rid of the `continue` added here.

This requires changing the method signature (this is a structural change and that should not be surprising). As passing in `job.get_hosts_for_fact_cache()` for hosts carries a non-trivial cost and it is no longer needed considering what you're doing.

Given that you don't need to pass it into `finish_fact_cache`, I'm also unclear if it should be passed into `start_fact_cache`. A semi-isolated way to do that might be to pass the full `Inventory` object from the outset.

I talked a good deal with @chrismeyersfsu about further consequences and I'm not sure if the refactor should even stop there. The contract between the start/finish methods is mainly the artifacts directory, and passing around the list of cached hosts in memory (given this particular kind of toxic code path) may be less maintainable than passing it around on disk. I would seriously consider saving the data structure in a file next to the fact cache folder (both being inside of artifacts). If you did that, I also expect it would lead into removing the `facts_write_time` argument.

djyasin replied:
@AlanCoding could you explain a little further what you mean by 

> Given that you don't need to pass it into finish_fact_cache, I'm also unclear if it should be passed into start_fact_cache. A semi-isolated way to do that might be to pass the full Inventory object from the outset.

Would you mind taking a look at what we have to see if it's good enough for this fix? I am happy to come up with a Jira to handle the additional refactoring if needed!

AlanCoding replied:
Left some much more minor comments. Yes, I can write a complete description of how I'd prefer it to work separately.

djyasin committed:
Fix formatting in awx/main/tests/unit/models/test_jobs.py

Rename for loop from host in hosts to hosts in hosts_cahced and remove unneeded continue

Revise finish_fact_cache to utilize inventory rather than hosts

Remove local var hosts that was assigned but unused

Revert change in start_fact_cache hosts_cached back to hosts

Revise the way we are handling hosts_cached and joining the file

Revert "Revise the way we are handling hosts_cached and joining the file"

This reverts commit e6e3d2f09c1b79a9bce3647a72e3dad97fe0aed8.

Reapply "Revise the way we are handling hosts_cached and joining the file"

This reverts commit a42b7ae69133fee24d3a5f1b456d9c343d111df9.

Revert some of my changes to get back to a better working state

Rename for loop to host in hosts_cached and remove unneeded continue

Remove jobs job.get_hosts_for_fact_cache() from post run hook, fix if statement after continue block, and revise how we are calling hosts in finish for loop

Add test_invalid_host_facts to test_jobs to increase code coverage

Update method signature to use hosts_cached and updated other references to hosts in finish_facts_cached

Rename hosts iterator to hosts_cached to agree with naming elsewhere

Revise test_invalid_host_facts to get more code coverage

Revise test_invalid_host_facts to increase codecov

Revise test_pre_post_run_hook_facts_deleted_sliced to ensure we are hitting the assertionerror for code cov

Revise  mock_inventory.hosts. to hit assert failure

Add revision of hosts and facts to force failure to satisfy code cov

Fix failure in test_pre_post_run_hook_facts_deleted_sliced

Add back for loop to create failures and add assert to hit them

Remove hosts.iterator() from both start_fact_cache and finish_fact_cache

Remove unused import of Queryset to satisfy api-lint

Fix typo in docstring hasnot to has not

Move hosts_cached.append(host) to outer loop in start_fact_cache

Add class to help support cached hosts resolving host.name issue with hosts_cached

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -36,9 +37,8 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
-
     last_filepath_written = None
-    for host in hosts:
+    for host in hosts_cached:
```
The `start_fact_cache` still needs to get the host list from the database, which is `job.get_hosts_for_fact_cache()`, which is still passed in as `hosts` to this method still. It's only the finishing method that should defer to the cached hosts from the starting method... somehow.

djyasin replied:
Thank you so much!! I've fixed that, this is very helpful!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -88,17 +89,17 @@ def update_hosts(host_list, max_tries=5):
-    if isinstance(hosts, QuerySet):
-        hosts = hosts.iterator()
+    if isinstance(hosts_cached, QuerySet):
+        hosts_cached = hosts_cached.iterator()
```
this `if` should be deleted, not modified. There is only 1 call we care about. With your changes, it only takes in what the `start_fact_cache` returns, which is a `list()` on creation and there's nothing that can ever change that later.

If it helps you, you could type annotate `host_cached: list` to make the assumptions behind this change more clear.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -34,9 +35,8 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
 
-    if isinstance(hosts, QuerySet):
+    if isinstance(hosts_cached, QuerySet):
         hosts = hosts.iterator()
```
Here, as well, can't be a `QuerySet` because you just made it a `list()`

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -130,6 +125,7 @@ def finish_fact_cache(hosts, destination, facts_write_time, log_data, job_id=Non
                 log_data['unmodified_ct'] += 1
         else:
             # if the file goes missing, ansible removed it (likely via clear_facts)
+            # if the file goes missing, but the host hasnot started facts, then we should not clear the facts
```
hasnot

AlanCoding reviewed (approved):
I filed AAP-43462 as the followup issue to this one.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -54,10 +53,13 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
             continue
-    # make note of the time we wrote the last file so we can check if any file changed later
+
+        hosts_cached.append(host)
```
This will miss any `host` entries that hit a `continue`, but your currently intended method requires that all "eligible" hosts are saved to this list.

That's a lot of hosts, and that's also why I would favor keeping `hosts_cached` as a list of _names_ as opposed to Django objects, as it is doing now.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -34,30 +42,32 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
                 json.dump(host.ansible_facts, f)
                 log_data['written_ct'] += 1
                 last_filepath_written = filepath
+            hosts_cached.append(CachedHost(name=host.name, inventory=host.inventory, ansible_facts=host.ansible_facts))
```
I still expect that this won't work. Again, when going through the `for host in hosts:` loop, many hosts will hit the `continue` statement before here, meaning they won't get into the list. Really, if you run this for the first time, **no** hosts will be in the list, and I legitimately think it will be `[]`. This is not what you want. For this to work as the variable of iteration in `finish_fact_cache`, it needs to basically contain all hosts in the inventory (of the section of inventory for sliced jobs).

And this is a very creative approach to create a new class, but keeping all ansible_facts in memory is too much. You also shouldn't need them, they shouldn't be used in `finish_fact_cache`.

AlanCoding committed:
Add live tests for ansible facts

Remove fixture needed for local work only maybe

Revert "Add class to help support cached hosts resolving host.name issue with hosts_cached"

This reverts commit 99d998cfb9960baafe887de80bd9b01af50513ec.

djyasin committed:
Move hosts_cached.append(host) outside of try except

djyasin committed:
Move hosts_cached.append(host) to the beginning of start_fact_cache

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -34,17 +34,17 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
     last_filepath_written = None
     for host in hosts:
-        if (not host.ansible_facts_modified) or (timeout and host.ansible_facts_modified < now() - datetime.timedelta(seconds=timeout)):
+        hosts_cached.append(host)
```
Everything else should be workable, but this would potentially demolish performance. We can pass a minimal set of attribues, like id & name, but the existing Ansible facts, specifically, will lead to excessive memory use.

tvo318 reviewed (approved):
Does this need a Yolo run attached to it?

djyasin committed:
Merge branch 'devel' into AAP_39365_Facts_are_unintentionally_deleted_when_the_inventory_is_modified_during_a_Job_execution
